rasterio/env.py

```diff
@@ -0,0 +1,167 @@
+    GDALEnv, del_gdal_config, get_gdal_config, set_gdal_config)
+from rasterio.dtypes import check_dtype
+from rasterio.errors import EnvError
+from rasterio.five import string_types
+from rasterio.transform import guard_transform
+from rasterio.vfs import parse_path, vsi_path
+
+
+# The currently active GDAL/AWS environment is a private attribute.
+_env = None
+
+log = logging.getLogger(__name__)
+
+
+class Env(object):
+    """Abstraction for GDAL and AWS configuration
+
+    The GDAL library is stateful: it has a registry of format drivers,
+    an error stack, and dozens of configuration options.
+
+    Rasterio's approach to working with GDAL is to wrap all the state
+    up using a Python context manager (see PEP 343,
+    https://www.python.org/dev/peps/pep-0343/). When the context is
+    entered GDAL drivers are registered, error handlers are
+    configured, and configuration options are set. When the context
+    is exited, drivers are removed from the registry and other
+    configurations are removed.
+
+    Example:
+
+        with Env(GDAL_CACHEMAX=512) as env:
+            # All drivers are registered, GDAL's raster block cache
+            # size is set to 512MB.
+            # Commence processing...
+            ...
+            # End of processing.
+
+        # At this point, configuration options are set to their
+        # previous (possible unset) values.
+
+    A boto3 session or boto3 session constructor arguments
+    `aws_access_key_id`, `aws_secret_access_key`, `aws_session_token`
+    may be passed to Env's constructor. In the latter case, a session
+    will be created as soon as needed. AWS credentials are configured
+    for GDAL as needed.
+    """
+
+    def __init__(self, aws_session=None, aws_access_key_id=None,
+                 aws_secret_access_key=None, aws_session_token=None,
+                 region_name=None, profile_name=None, **options):
+        """Create a new GDAL/AWS environment.
+
+        Note: this class is a context manager. GDAL isn't configured
+        until the context is entered via `with Env():`
+
+        Parameters
+        ----------
+        aws_session: object, optional
+            A boto3 session.
+        **options: optional
+            A mapping of boto3 session constructor keyword arguments
+            and GDAL configuration options.
+
+        Returns
+        -------
+        A new instance of Env.
+        """
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.aws_session_token = aws_session_token
+        self.region_name = region_name
+        self.profile_name = profile_name
+        self.aws_session = aws_session
+        self._creds = (
+            self.aws_session._session.get_credentials()
+            if self.aws_session else None)
+        self.options = options.copy()
+        self.previous_options = {}
+        defenv()
+
+    def get_aws_credentials(self):
```

Comment:
@sgillies The AWS credentials are not pushed into the GDAL environment until `Env.get_aws_credentials()` is called.

```python
import rasterio as rio
from rasterio.env import Env
from rasterio import env


with Env(aws_access_key_id='whatever', CPL_DEBUG=True) as e:
    print(env.getenv())
    e.get_aws_credentials()
    print(env.getenv())
```

results:

```
{'CPL_DEBUG': True}
{'aws_access_key_id': 'whatever', 'CPL_DEBUG': True}
```

It looks like this might cause a problem when instantiating new `Env()`'s that should be inheriting modified credentials from an outer `Env()`, like https://github.com/mapbox/rasterio/pull/665#discussion_r60863582.